zebROS_ws/src/controllers_2023/src/elevator_controller.cpp

```diff
@@ -0,0 +1,387 @@
+	last_zeroed_  = false;
+	last_time_down_ = ros::Time::now();
+	last_mode_ = hardware_interface::TalonMode_Disabled;
+	last_position_ = -1; // give nonsense position to force update on first time through update()
+	position_command_.writeFromNonRT(ElevatorCommand_2023());
+}
+
+void ElevatorController_2023::update(const ros::Time &/*time*/, const ros::Duration &/*duration*/)
+{
+	// If we hit the limit switch, (re)zero the position.
+	if (elevator_joint_.getReverseLimitSwitch())
+	{
+		ROS_INFO_THROTTLE(2, "ElevatorController_2023 : hit limit switch");
+		if (!last_zeroed_)
+		{
+			zeroed_ = true;
+			last_zeroed_ = true;
+			elevator_joint_.setSelectedSensorPosition(0);
+			elevator_joint_.setDemand1Type(hardware_interface::DemandType_ArbitraryFeedForward);
+			elevator_joint_.setDemand1Value(config_.arb_feed_forward_low);
+		}
+	}
+	else
+	{
+		last_zeroed_ = false;
+	}
+
+
+	if (zeroed_) // run normally, seeking to various positions
+	{
+		elevator_joint_.setMode(hardware_interface::TalonMode_MotionMagic);
+		if (((elevator_joint_.getMode()) == (hardware_interface::TalonMode_Disabled)) && ((last_mode_) != (hardware_interface::TalonMode_Disabled)))
+		{
+			position_command_.writeFromNonRT(ElevatorCommand (elevator_joint_.getPosition()));
+		}
+		const ElevatorCommand_2023 setpoint = *(position_command_.readFromRT());
+		elevator_joint_.setCommand(setpoint.GetPosition());
+
+		//if we're not climbing, add an arbitrary feed forward to hold the elevator up
+		if(!setpoint.GetGoSlow())
+		{
+			elevator_joint_.setMotionAcceleration(config_.motion_magic_acceleration);
+			elevator_joint_.setMotionCruiseVelocity(config_.motion_magic_velocity);
+			elevator_joint_.setPIDFSlot(0);
+			// Add arbitrary feed forward for upwards motion
+			// We could have arb ff for both up and down, but seems
+			// easier (and good enough) to tune PID for down motion
+			// and add an arb FF correction for up
+
+			if(elevator_joint_.getPosition() >= config_.stage_2_height && last_position_ <= config_.stage_2_height) {
+				elevator_joint_.setDemand1Type(hardware_interface::DemandType_ArbitraryFeedForward);
+				elevator_joint_.setDemand1Value(config_.arb_feed_forward_high);
+			}
+			else if (elevator_joint_.getPosition() <= config_.stage_2_height && last_position_ >= config_.stage_2_height) {
+				elevator_joint_.setDemand1Type(hardware_interface::DemandType_ArbitraryFeedForward);
+				elevator_joint_.setDemand1Value(config_.arb_feed_forward_low);
+			}
+
+
+			//for now, up and down PID is the same, so slot 1 is used for climbing
+			/*
+			if(last_setpoint_ != setpoint) {
+				if(setpoint > elevator_joint_.getPosition()) {
+					elevator_joint_.setPIDFSlot(0);
+				}
+				else {
+					elevator_joint_.setPIDFSlot(1);
+				}
+			}
+			last_setpoint_ = setpoint;
+			*/
+		}
+		
+	}
+	else
+	{
+		elevator_joint_.setMode(hardware_interface::TalonMode_PercentOutput);
+		if ((ros::Time::now() - last_time_down_).toSec() < config_.elevator_zeroing_timeout)
```

Comment:
I probably forgot this one also - the update call has access to the current time in the time function argument. Let's use that one 
- uncomment the time function argument 
- use that time argument instead of calling ros::Time::now()